xde/src/xde.rs

```diff
@@ -2059,27 +2075,53 @@ unsafe extern "C" fn xde_rx(
     // dropped yet and thus our `MacClientHandle` is also still valid.
     let mch_ptr = arg as *const MacClientHandle;
     Arc::increment_strong_count(mch_ptr);
-    let mch = Arc::from_raw(mch_ptr);
+    let mch: Arc<MacClientHandle> = Arc::from_raw(mch_ptr);
```

Comment:
Vague vibes:

It hasn't changed in this PR, but it makes nervous to see `Arc::from_raw` paired with `Arc::as_ptr` (instead of `Arc::into_raw`).

I read through the `MacClientHandle::add_promisc` and think that it's correct as written.  However, I wonder if it would be cleaner to use `Arc::into_raw` and store a `*const MacClientHandle` in `MacPromiscHandle` (then rebuild the `Arc` in `Drop`), e.g. something like.

```rust
impl MacClientHandle {
    // ...

    /// Register promiscuous callback to receive packets on the underlying MAC.
    pub fn add_promisc(
        self: &Arc<Self>,
        ptype: mac_client_promisc_type_t,
        promisc_fn: mac_rx_fn,
        flags: u16,
    ) -> Result<MacPromiscHandle, c_int> {
        let mut mph = ptr::null_mut();

        // `MacPromiscHandle` keeps a reference to this `MacClientHandle`
        // in `mch` until it is removed and so we can safely access it from the
        // callback via the `arg` pointer.
        let mch = Arc::into_raw(self.clone());
        let ret = unsafe {
            mac_promisc_add(self.mch, ptype, promisc_fn, mch as *mut c_void, &mut mph, flags)
        };

        if ret == 0 {
            Ok(MacPromiscHandle { mph, mch })
        } else {
            Err(ret)
        }
    }
// ...
}

/// Safe wrapper around a `mac_promisc_handle_t`.
#[derive(Debug)]
pub struct MacPromiscHandle {
    /// The underlying `mac_promisc_handle_t`.
    mph: *mut mac_promisc_handle,

    /// The `MacClientHandle` used to create this promiscuous callback.
    mch: *const MacClientHandle,
}

impl Drop for MacPromiscHandle {
    fn drop(&mut self) {
        // Safety: We know that a `MacPromiscHandle` can only exist if a
        // mac promisc handle was successfully obtained, and thus `mph`
        // is valid.
        unsafe {
            mac_promisc_remove(self.mph);
            Arc::from_raw(self.mch); // dropped immediately
        };
    }
}
```

Reply:
I've tested this out and it seems to behave well, thanks for the suggestion!